src/mlpack/methods/pca/decomposition_policies/stochastic_method.hpp

```diff
@@ -0,0 +1,287 @@
+     *
+     * @param coordinates The function coordinates.
+     * @param begin The first function.
+     * @param gradient The function gradient.
+     * @param batchSize Number of points to process.
+     */
+    void Gradient(const arma::mat& coordinates,
+                  const size_t begin,
+                  arma::mat& gradient,
+                  const size_t batchSize)
+    {
+      gradient = data.cols(begin, begin + batchSize - 1) *
+          data.cols(begin, begin + batchSize - 1).t() * coordinates;
+    }
+
+    /*
+     * Evaluate the objective function and gradient of the PCA function for a
+     * particular batch-size.
+     *
+     * @param coordinates The function coordinates.
+     * @param begin The first function.
+     * @param gradient The function gradient.
+     * @param batchSize Number of points to process.
+     */
+    double EvaluateWithGradient(const arma::mat& coordinates,
+                                const size_t begin,
+                                arma::mat& gradient,
+                                const size_t batchSize)
+    {
+      gradient = data.cols(begin, begin + batchSize - 1) *
+          data.cols(begin, begin + batchSize - 1).t() * coordinates;
+
+      return arma::trace(coordinates * gradient);
+    }
+
+   private:
+    //! Data matrix..
+    arma::mat data;
+  };
+};
+
+/**
+ * Implementation of the PCA Update policy which wraps the existing optimizer
+ * update policies.
+ */
+template<typename UpdatePolicyType = optimization::VanillaUpdate>
+class PCAUpdate
+{
+ public:
+  /**
+   * Construct the wrapper update policy with the given update policy.
+   *
+   * @param beta1 Exponential decay rate for the first moment estimates.
+   * @param beta2 Exponential decay rate for the weighted infinity norm
+   *        estimates.
+   * @param epsilon Value used to initialise the mean squared gradient
+   *        parameter.
+   */
+  PCAUpdate(const UpdatePolicyType& updatePolicy = UpdatePolicyType()) :
+      updatePolicy(updatePolicy)
+  {
+    /* Nothing to do here. */
+  }
+
+  /**
+   * Adam specific constructor.
+   *
+   * @param beta1 Exponential decay rate for the first moment estimates.
+   * @param beta2 Exponential decay rate for the weighted infinity norm
+   *        estimates.
+   * @param epsilon Value used to initialise the mean squared gradient
+   *        parameter.
+   */
+  PCAUpdate(const double epsilon, const double beta1, const double beta2) :
+      updatePolicy(UpdatePolicyType(epsilon, beta1, beta2))
+  {
+    /* Nothing to do here. */
+  }
+
+  void Initialize(const size_t rows, const size_t cols)
+  {
+    updatePolicy.Initialize(rows, cols);
+  }
+
+  /**
+   * General update step, all parameters are forwarded to the actual
+   * update policy.
+   *
+   * @param iterate Parameters that minimize the function.
+   */
+  template<typename... Targs>
+  void Update(arma::mat& iterate, Targs... fArgs)
+  {
+    iterate *= -1;
```

Comment:
(Oops, I put this in the wrong place originally.)

I guess, I am still just confused about this bit---shouldn't this be unnecessary if we simply negate the results of Evaluate() and EvaluateWithGradient() and Gradient()? Sorry if I am overlooking something---it's possible I've misunderstood. It just seems like extra computation here.

Reply:
Ah, sorry that I never responded to this.  I follow now, sorry for the confusion.